h/accounts/services.py

```diff
@@ -47,19 +48,41 @@ def __init__(self, default_authority, session):
         def flush_cache(session):
             self._cache = {}
 
-    def fetch(self, userid):
+    def fetch(self, userid_or_username, authority=None):
         """
-        Fetch a user by userid, e.g. 'acct:foo@example.com'
+        Fetch a user by userid or by username and authority.
+
+        Takes *either* a userid *or* a username and authority as arguments.
+        For example::
+
+          user_service.fetch('acct:foo@example.com')
+
+        or::
+
+          user_service.fetch('foo', 'example.com')
 
         :returns: a user instance, if found
         :rtype: h.models.User or None
+
         """
-        if userid not in self._cache:
-            self._cache[userid] = (self.session.query(User)
-                                   .filter_by(userid=userid)
-                                   .one_or_none())
+        if authority is not None:
+            username = userid_or_username
+        else:
+            userid = userid_or_username
+            parts = util.user.split_user(userid)
+            username = parts['username']
+            authority = parts['domain']
+
+        # The cache is keyed by (username, authority) tuples.
```

Comment:
It would be helpful to include this documentation at the point where `self._cache` is "declared" (since you don't pre-declare fields in Python, the point of first initialization in the constructor is a good alternative)